source/wrangling.md

```diff
@@ -1152,1262 +974,807 @@ glue("census_popn", "{0:,.0f}".format(35151728))
-```
+:tags: [remove-cell]
 
-```{code-cell} ipython3
-region_lang.iloc[:, 3:]
+english_lang = region_lang[region_lang["language"] == "English"]
+five_cities = ["Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton"]
+english_lang = english_lang[english_lang["region"].isin(five_cities)]
+english_lang
 ```
 
-Now we use `.apply` with argument `axis=1`, to tell Python that we would like
-the `max` function to be applied across, and within, a row,
-as opposed to being applied on a column 
-(which is the default behavior of `.apply`):
-
+Sometimes you want to create a new data frame. You can use `assign` to create a data frame from scratch.
+Lets return to the example of wanting to compute the proportions of people who speak English
+most at home in Toronto, Montréal, Vancouver, Calgary, Edmonton. Before adding new columns, we filtered
+our `region_lang` to create the `english_lang` data frame containing only English speakers in the five cities
+of interest.
 ```{code-cell} ipython3
----
-jupyter:
-  source_hidden: true
-tags: [remove-cell]
----
-# Now we apply `rowwise` before `mutate`, to tell R that we would like
-# the mutate function to be applied across, and within, a row,
-# as opposed to being applied on a column 
-# (which is the default behavior of `mutate`):
+:tags: ["output_scroll"]
+english_lang
 ```
+We then wanted to add the populations of these cities as a column using `assign`
+(Toronto: 5928040, Montréal: 4098927, Vancouver: 2463431,
+Calgary: 1392609, and Edmonton: 1321426). We had to be careful to add those populations in the
+right order, and it could be easy to make a mistake this way. An alternative approach, that we demonstrate here
+is to (1) create a new, empty data frame, (2) use `assign` to assign the city names and populations in that
+data frame, and (3) use `merge` to combine the two data frames, recognizing that the "regions" are the same.
 
+We start by creating a new, empty data frame by calling `pd.DataFrame` and assigning the output
+to `city_populations`. We then use `assign` to add the city names in a column called `"region"`
+and their populations in a column called `"population"`.
 ```{code-cell} ipython3
-region_lang_rowwise = region_lang.assign(
-    maximum=region_lang.iloc[:, 3:].apply(max, axis=1)
+city_populations = pd.DataFrame()  # empty data frame
+city_populations = city_populations.assign(
+  region=["Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton"],
+  population=[5928040, 4098927, 2463431, 1392609, 1321426]
 )
-
-region_lang_rowwise
+city_populations
 ```
-
-We see that we get an additional column added to the data frame, 
-named `maximum`, which is the maximum value between `mother_tongue`,
-`most_at_home`, `most_at_work` and `lang_known` for each language
-and region.
-
+This new data frame has the same `region` column as the `english_lang` data frame. The order of
+the cities is different, but that is okay, we can use the `merge` function in `pandas` to say
```

Comment:
note that I added the introduction of `merge` here. It seems like a natural example but was not in the previous version, so please see what you think

Reply:
more wrangling more better!